src/ut/alarm_heap_test.cpp

```diff
@@ -0,0 +1,362 @@
+
+inline Matcher<netsnmp_variable_list*> TrapVars(TrapVarsMatcher::TrapType trap_type,
+                                                unsigned int alarm_index) {
+  return MakeMatcher(new TrapVarsMatcher(trap_type, alarm_index));
+}
+
+// Simple test that raising an alarm triggers an INFORM to be sent immediately
+TEST_F(AlarmHeapTest, SetAlarm)
+{
+  COLLECT_CALL(send_v2trap(TrapVars(TrapVarsMatcher::ACTIVE, 1000), _, _));
+  _alarm_heap->issue_alarm("test", "1000.3");
+  _ms.trap_complete(1, 5);
+}
+
+// Simple test that clearing an alarm triggers an INFORM to be sent immediately
+// (from a clean start).
+TEST_F(AlarmHeapTest, ClearAlarm)
+{
+  COLLECT_CALL(send_v2trap(TrapVars(TrapVarsMatcher::CLEAR, 1000), _, _));
+  _alarm_heap->issue_alarm("test", "1000.1");
+  _ms.trap_complete(1, 5);
+}
+
+// Test that repeated alarms only generate one INFORM
+TEST_F(AlarmHeapTest, SetAlarmRepeatedState)
+{
+  COLLECT_CALL(send_v2trap(TrapVars(TrapVarsMatcher::ACTIVE, 1000), _, _));
+  _alarm_heap->issue_alarm("test", "1000.3");
+  _alarm_heap->issue_alarm("test", "1000.3");
+  _ms.trap_complete(1, 5);
+}
+
+// Test that increasing alarm severity generates INFORMs immediately
+TEST_F(AlarmHeapTest, SetMultiAlarmIncreasingState)
+{
+  COLLECT_CALL(send_v2trap(TrapVars(TrapVarsMatcher::CLEAR, 2000), _, _));
+  _alarm_heap->issue_alarm("test", "2000.1");
+  _ms.trap_complete(1, 5);
+
+  COLLECT_CALL(send_v2trap(TrapVars(TrapVarsMatcher::ACTIVE, 2000), _, _));
+  _alarm_heap->issue_alarm("test", "2000.5");
+  _ms.trap_complete(1, 5);
+
+  COLLECT_CALL(send_v2trap(TrapVars(TrapVarsMatcher::ACTIVE, 2000), _, _));
+  _alarm_heap->issue_alarm("test", "2000.4");
+  _ms.trap_complete(1, 5);
+}
+
+
+// Test that syncing alarms generates INFORMs for all known alarms immediately
+TEST_F(AlarmHeapTest, SyncAlarms)
+{
+  // Put our three alarms into a state we expect
+  COLLECT_CALL(send_v2trap(TrapVars(TrapVarsMatcher::ACTIVE,
+                                        1000), _, _));
+
+  COLLECT_CALL(send_v2trap(TrapVars(TrapVarsMatcher::CLEAR,
+                                        1001), _, _));
+
+  COLLECT_CALL(send_v2trap(TrapVars(TrapVarsMatcher::CLEAR,
+                                        1002), _, _));
+
+  _alarm_heap->issue_alarm("test", "1000.3");
+  _alarm_heap->issue_alarm("test", "1001.1");
+  _alarm_heap->issue_alarm("test", "1002.1");
+
+  _ms.trap_complete(3, 5);
+
+  // Now call sync_alarms and expect those states to be retransmitted
+  COLLECT_CALL(send_v2trap(TrapVars(TrapVarsMatcher::ACTIVE,
+                                        1000), _, _));
+
+  COLLECT_CALL(send_v2trap(TrapVars(TrapVarsMatcher::CLEAR,
+                                        1001), _, _));
+
+  COLLECT_CALL(send_v2trap(TrapVars(TrapVarsMatcher::CLEAR,
+                                        1002), _, _));
+
+  _alarm_heap->sync_alarms();
+
+  _ms.trap_complete(3, 5);
+}
+
+// Test that an alarm flicker situation doesn't cause flickering INFORMs
```

Comment:
I can't immediately see what about this test would fail if it did flicker - worth a clarifying comment?